Add support for grouping source map loading errors.

jdatapple opened the pull request:
#### 869419c2de211c08fe24d0077a09d5289c06dc4f
<pre>
Add support for grouping source map loading errors.
<a href="https://bugs.webkit.org/show_bug.cgi?id=256683">https://bugs.webkit.org/show_bug.cgi?id=256683</a>
<a href="https://rdar.apple.com/problem/109239646">rdar://problem/109239646</a>

Reviewed by Devin Rousso.

This feature is behind an experimental Web Inspector setting
available in the Settings tab, under Experimental. When enabled and
more than one source map loading error occurs, network error messages
with a URL that matches a known source map URL are collected into an
expandable group with a counter showing the number of messages.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager):
(WI.ConsoleManager.prototype.get failedSourceMapConsoleMessages):
(WI.ConsoleManager.prototype.messageWasAdded):
(WI.ConsoleManager.prototype._handleSourceMapErrorMessage):
* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
(WI.JavaScriptLogViewController):
(WI.JavaScriptLogViewController.prototype.startNewSession):
(WI.JavaScriptLogViewController.prototype._didRenderConsoleMessageView):
(WI.JavaScriptLogViewController.prototype._handleFailedSourceMapMessageGrouping):
* Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
(WI.NetworkManager.prototype.get failedSourceMapURLs):
(WI.NetworkManager.prototype.get downloadingSourceMaps):
(WI.NetworkManager.prototype.isDownloadingSourceMapURL):
(WI.NetworkManager.prototype._sourceMapLoadAndParseFailed):
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.css:
(.console-sourcemap-errors-group .errors-count,):
(.console-message .console-sourcemap-failures-container,):
(.console-message.expandable.expanded .console-sourcemap-failures-container,):
(.console-message .repeat-count): Deleted.
(.console-message .console-message-extra-parameters-container,): Deleted.
(.console-message.expandable.expanded .console-message-extra-parameters-container,): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js: Added.
(WI.ConsoleSourceMapMessageGroup):
(WI.ConsoleSourceMapMessageGroup.prototype.render):
(WI.ConsoleSourceMapMessageGroup.prototype.addMessageView):
(WI.ConsoleSourceMapMessageGroup.prototype.toggle):
(WI.ConsoleSourceMapMessageGroup.prototype._renderMessageCount):
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:

Canonical link: <a href="https://commits.webkit.org/270834@main">https://commits.webkit.org/270834@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/216389df241d9419339236cac0e6fef3cd85e81a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5181 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28778 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24296 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6989 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2588 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26826 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/4011 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/22824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3532 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3574 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/23810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29260 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24225 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/24209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3610 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27736 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5046 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6371 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/4084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3938 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +373,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
                 this._currentSessionOrGroup.append(groupElement);
                 this._currentSessionOrGroup = group;
-            } else
+            } else if (WI.consoleManager.failedSourceMapConsoleMessages?.has(messageView.message)) {
```
No need for `?` in `failedSourceMapConsoleMessages?.has()`. 
The getter always exists and returns a `Set` created in the constructor of `ConsoleManager`

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -293,6 +297,17 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
         WI.objectStores.consoleSnippets.putObject(snippet);
     }
 
+    _handleSourceMapErrorMessage(message)
```
Web Inspector generally uses methods prefixed with `_handle` for event handlers. 
Is there a different name that can be used?

jdatapple replied:
I went with `_maybeAddConsoleMessageToConsoleSourceMapErrorGroup` to be clear about the effect.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +373,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
             this.delegate.didAppendConsoleMessageView(messageView);
     }
+    
+    _handleFailedSourceMapMessageGrouping(messageView)
```
`_handle`-prefixed methods are generally used to identify an event handler.  

How about renaming `_handleFailedSourceMapMessageGrouping` to `_addConsoleMessageToFailedSourceMapsGroup` or something similar that explains the intent?

jdatapple replied:
I used `_addConsoleMessageToConsoleSourceMapErrorGroup` here.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +373,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+
+        // Create the group, add the first tracked message view
+        if (this._failedSourceMapsGroup === null) {
+            this._failedSourceMapsGroup = new WI.ConsoleSourceMapMessageGroup(this._currentSessionOrGroup);
```
`this._currentSessionOrGroup` may point to another `WI.ConsoleGroup` instead of a top-level `WI.ConsoleSession`. 

That can happen if a `console.group()` was executed before the first `WI.ConsoleMessage` for a failed source map request comes in and a `console.groupEnd()` wasn't yet executed. 

This would nest `WI.ConsoleSourceMapMessageGroup` under a `WI.ConsoleGroup` which we probably don't want. 

You'll likely need to hold a reference to the current `WI.ConsoleSession` when it gets instantiated (`this._currentSession`, perhaps?) and append `WI.ConsoleSourceMapMessageGroup` to that.

And this makes me wonder what happens when there are multiple sessions, for example when using the "Preserve Log Messages" option in the Console to keep old messages when reloading the inspected page. Check to ensure that console messages for failed resource map requests don't go into a group associated with a previous session.

jdatapple replied:
The update uses `let currentSession = this._sessions.lastValue`.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +373,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+        if (!WI.settings.experimentalGroupSourceMapErrors.value)
+            return;
+
+        // Create the group, add the first tracked message view
```
The code comment is redundant with the code. It can be removed.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +373,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+            this._currentSessionOrGroup.append(groupElement);
+        }
+
+        // Add all subsequent message views to the group
```
The code comment is misleading (the line adds the current message to the group) but also redundant. It can be removed.

jdatapple replied:
Removed it.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +373,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+        if (this._failedSourceMapsGroup === null) {
+            this._failedSourceMapsGroup = new WI.ConsoleSourceMapMessageGroup(this._currentSessionOrGroup);
+
+            let groupElement = this._failedSourceMapsGroup.render();
```
If `WI.ConsoleSourceMapMessageGroup.render()` _needs_ to be called for the component to be useful and it doesn't need an argument, you can call it automatically in `WI.ConsoleSourceMapMessageGroup`\`s constructor and remove the call to `render()`.

Then, you can expose the container element via `WI.ConsoleSourceMapMessageGroup.element`

jdatapple replied:
I moved this to the `WI.ConsoleSourceMapMessageGroup` constructor as recommended.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,124 @@
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+WI.ConsoleSourceMapMessageGroup = class ConsoleSourceMapMessageGroup extends WI.ConsoleGroup
```
There's enough divergence from `WI.ConsoleGroup` and none of its behaviors as a subclass are used so it probably doesn't warrant to have `WI.ConsoleSourceMapMessageGroup` extend it. 

Perhaps this can be a standalone class extending `WI.Object`.

jdatapple replied:
As we discussed, this new approach reuses as much of `WI.ConsoleGroup` as possible.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,124 @@
+    {
+        var groupElement = document.createElement("div");
+        groupElement.classList.add("console-sourcemap-errors-group");
+        groupElement.group = this;
```
`groupElement.group` doesn't seem to be used anywhere.

jdatapple replied:
Removed from here, and also from `WI.ConsoleGroup` as a drive-by improvement.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,124 @@
+        groupElement.group = this;
+        this.element = groupElement;
+
+        // TODO set consoleMessage to use WI.UIStringWI.UIString("Source Map loading errors")
```
The TODO comment is obsolete.

jdatapple replied:
Yep. 🤦

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,124 @@
+        consoleMessageView.render();
+
+        var titleElement = consoleMessageView.element;
+        this._titleElement = titleElement;
```
This can be inlined: `this._titleElement = consoleMessageView.element`

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,124 @@
+        this._titleElement._expandable = false;
+
+        this._boundClickHandler = this.toggle.bind(this);
+        this._titleElement.addEventListener("click", this._boundClickHandler);
```
Style: Web Inspector prefers arrow functions instead of bound functions for event handlers:


```
this._titleElement.addEventListener("click", () => { this.toggle(); });
```

jdatapple replied:
Addressed in the new approach.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,124 @@
+        if (this._titleElement.classList.contains("expanded"))
+            this._titleElement.classList.remove("expanded");
+        else
+            this._titleElement.classList.add("expanded");
```
`this._titleElement.classList.toggle("expanded")` ? In which case, you can likely handle it inline when adding the "click" event listener.

jdatapple replied:
This is all removed and handled by `WI.ConsoleGroup` behavior.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,124 @@
+        element.classList.add(WI.LogContentView.ItemWrapperStyleClassName);
+
+        let messageBodyElement = element.getElementsByClassName("console-message-body")?.item(0);
+        messageBodyElement?.classList.remove("console-top-level-message");
```
Why does the `"console-top-level-message"` classname need to be removed from the incoming `ConsoleMessageView`?

Nesting a message into group shouldn't have to mutate its DOM, the same way as `ConsoleGroup.addMessageView()` does not.

Even if it were need, we generally prefer to have a method on `ConsoleMessageView` that does that instead of poking inside its DOM via `querySelector` or similar methods.

jdatapple replied:
As discussed, none of my conclusions about `"console-top-level-message"` were correct and I've avoided it all in the new approach.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,124 @@
+            this._titleElement.classList.remove("hidden");
+            this._messagesElement.classList.add("console-sourcemap-failures-container");
+            this._titleElement.appendChild(this._messagesElement);
+        }
```
I think I understand what you're trying to do here (mutate the group's DOM if it only has one `ConsoleMessageView` so it's not collapsible / expandable), but I'll suggest an alternative.

A group should always be collapsible. Keeping its internals simple also helps readers later on.

I suggest keeping a reference to the first `ConsoleMessageView` for a failed source map request on `WI.JavaScriptLogViewController` (i.e.: append the first message in  `WI.JavaScriptLogViewController._currentSessionOrGroup` as with other messages).

Once there's a second `ConsoleMessageView` for a failed source map request:
- create the `WI.ConsoleSourceMapMessageGroup`
- ~append its element _before_ the first `ConsoleMessageView.element`~
- **EDIT**: append its element to the `_currentSessionOrGroup` (inserting right before/after the first `ConsoleMessageView` isn't strictly necessary since we'll be moving the first message into the group anyway. `_currentSessionOrGroup` can be a `ConsoleGroup` or `ConsoleSession`; you'd have to expose their respective DOM `element` in order to use `insertBefore()`)
- reparent that first `ConsoleMessageView` into the group (the signature of `ConsoleGroup.addMessageView()` is indicative of this intent)
- append the second `ConsoleMessageView` into the group. 
- continue appending subsequent messages into the group (i.e. your current patch)

This would obviate the need for CSS class name gymnastics inside the group and I believe would simplify the code make it easier to follow.

Something along these lines:

```diff
_handleFailedSourceMapMessageGrouping(messageView)
    {
        // ...
         
+        if (!this._firstFailedSourceMapConsoleMessageView) {
+          this._currentSessionOrGroup.addMessageView(messageView);
+          this._firstFailedSourceMapConsoleMessageView = messageView;
+          return;
+       }
         
        // Create the group, add the first tracked message view
        if (this._failedSourceMapsGroup === null) {
            this._failedSourceMapsGroup = new WI.ConsoleSourceMapMessageGroup(this._currentSessionOrGroup);

            let groupElement = this._failedSourceMapsGroup.render();
            this._currentSessionOrGroup.append(groupElement);
+          this._failedSourceMapsGroup.addMessageView(this._firstFailedSourceMapConsoleMessageView);
        }

        // Add all subsequent message views to the group
        this._failedSourceMapsGroup.addMessageView(messageView);
    }
```

jdatapple replied:
Addressed in the new approach.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +373,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+    _handleFailedSourceMapMessageGrouping(messageView)
+    {
+        if (!WI.settings.experimentalGroupSourceMapErrors.value)
+            return;
```
If the setting is false, console messages for failed source map requests will not be added to the view.

The `else` branch in `_didRenderConsoleMessageView()` won't be reached. 

Either move this setting check to the `if` statement outside or include the alternative operation here before early returning: `this._currentSessionOrGroup.addMessageView(messageView);`

jdatapple replied:
This was addresed by moving the check to the upper `if` statement.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,124 @@
+            this._titleElement.insertBefore(this._messageCountElement, this._titleElement.firstChild);
+        }
+
+        this._messageCountElement.textContent = Number.abbreviate(count);
```
If opting to do the reparenting of the first console message into an always collapsible group, 
the `this._messageCountElement` can be created in the constructor (or in `render()`) and the count can be incremented in `addMessageView`, thus obviating the need for this entire `_renderMessageCount()` method.

jdatapple replied:
Addressed this in the new approach.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,124 @@
+        var titleElement = consoleMessageView.element;
+        this._titleElement = titleElement;
+        this._titleElement.classList.add(WI.LogContentView.ItemWrapperStyleClassName, "hidden");
+        this._titleElement._expandable = false;
```
I commented with an alternative to always keep the group collapsible / expandable. 
But if we _needed_ to keep this state, it should live on a member of class not on a DOM element.

jdatapple replied:
Addressed in the new approach.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
```diff
@@ -328,6 +337,11 @@ WI.NetworkManager = class NetworkManager extends WI.Object
         loadAndParseSourceMap();
     }
 
+    isDownloadingSourceMapURL(url)
```
This method is accessing URLs in two different buckets, despite its name. 
Since the intent is to know whether the URL is for a source map url, regardless of its state, how about renaming to `isSourceMapURL(url)`?

jdatapple replied:
I did change to `isSourceMapURL(url)`, but there is a `Map` on the `networkManager` called `_sourceMapURLMap` that has all of the identified source map URLs. This raises a question of whether it is better to reference that instead, or continue to only look at URLs where a download attempt has occurred, or a download failed. The correct answer still seems like the latter, but I think either approach will have the same behavior result in practice.

rcaliman-apple reviewed (changes requested):
Nice work! I like the approach. 👍 

I left some comments about some edge cases and a suggestion that removes the need to mutate the DOM of the group for one or more messages. Some style and nits as well.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,61 @@
+        errorIconElement.classList.add("console-error-level");
+        this._titleConsoleMessageView.element.insertBefore(errorIconElement, this._titleConsoleMessageView.element.firstChild);
+
+        this.element.classList.add("console-group-sourcemap-errors");
```
There's an error accessing `this.element` at this point because it gets assigned in `ConsoleGroup.render()` which is only called on the next line.

But I have another suggestion that will likely make this unnecessary.

jdatapple replied:
Addressed in deba5de495d803c6f9703282319fb8296e21b428.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleGroup.js:
```diff
@@ -47,7 +47,6 @@ WI.ConsoleGroup = class ConsoleGroup extends WI.Object
         var groupElement = document.createElement("div");
         groupElement.className = "console-group";
-        groupElement.group = this;
         this.element = groupElement;
```
Drive-by since you're changing this file: please assign to `this._element` and add an explicit getter to `get element { return this._element; }` to follow convention in Web Inspector. 

It was probably an oversight that this was directly assigned and exposed.

jdatapple replied:
Ahh, I see. Addressed in deba5de495d803c6f9703282319fb8296e21b428.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,61 @@
+        
+        let errorIconElement = document.createElement("span");
+        errorIconElement.classList.add("console-error-level");
+        this._titleConsoleMessageView.element.insertBefore(errorIconElement, this._titleConsoleMessageView.element.firstChild);
```
The UI convention in Web Inspector is for the error icon to appear before the disclosure toggle and count.

Current patch:

<img width="440" alt="Screenshot 2023-10-11 at 23 00 07" src="https://github.com/WebKit/WebKit/assets/105809798/541a122c-f7b7-4eb2-a77e-8baa0d714b55">

A screenshot of repeat errors:

<img width="295" alt="Screenshot 2023-10-11 at 21 06 16" src="https://github.com/WebKit/WebKit/assets/105809798/7dcc68c5-83c6-45d3-849a-88c07b9c5b8a">

To achieve this, instead of generating a `<span>`  and injecting it just to hold the icon, we could reuse the `.console-error-level` CSS class and assign it directly to the console group container, i.e.: the element that matches `.console-group`.

This means we can probably generalize and remove the need for the new `.console-group-sourcemap-errors` CSS class name:

```diff
- <div class="console-group console-group-sourcemap-errors">
+ <div class="console-group console-error-level">
```

To account for the fact that that will add an over-sized error icon on the `::before` pseudo-element, we can mitigate by extending (or copying) the CSS rule with boilerplate for icons from `LogContentView.css`:

```diff
.console-item::before,
+ .console-group::before {
    position: absolute;
    left: 4px;
    height: 12px;
    width: 12px;
+  z-index: 1;
}
```

A few more tweaks are needed:

```diff
+.console-group::before {
+    top: 4px; // because .console-group does not have a top/bottom padding like .console-message
+}
```

Then, we need to make room for the two icons side-by-side (error and disclosure triangle), so we add some inline padding:

```diff
+.console-group:is(.console-warning-level, .console-error-level, .console-log-level, .console-info-level, .console-debug-level) {
+    padding-inline-start: 16px;
+}
```

This has the added benefit of indenting the whole contents of the console group, replacing the functionality of this other CSS rule:


```diff
-.console-group-sourcemap-errors .console-group-messages {
-    padding-left: 16px;
-}
```

I tried out the above in a debugging session and the UI looks similar to the one you propose (plus error icon repositioning):

<img width="472" alt="Screenshot 2023-10-11 at 21 42 38" src="https://github.com/WebKit/WebKit/assets/105809798/a5ade09d-9519-4ed8-8587-98e743c14f2a">

This seems viable. 

---

If you agree with this approach, I suggest extending `WI.ConsoleGroup` with something like `WI.ConsoleGroup.setMessageLevel()` which would set the appropriate level class name  on `WI.ConsoleGroup.element` (see [ConsoleMessageView.js#L70-L91](https://github.com/WebKit/WebKit/blob/92948d2597d6772eeb45cd85aec2af9a58167f37/Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js#L70-L91)).

For example:
```
consoleGroup.setMessageLevel(WI.ConsoleMessage.MessageLevel.Error)
```

The CSS related to `WI.ConsoleGroup` from `LogContentView.css` should likely be moved to a new `ConsoleGroup.css` stylesheet to reflect that `ConsoleGroup` is now a "component".


> [!NOTE]
> This suggestion essentially reuses the styling of message levels (error, info, warning, etc) on console groups without having to specialize the CSS for `WI.ConsoleSourceMapMessageGroup`

jdatapple replied:
This is a great suggestion. I did an implementation in deba5de495d803c6f9703282319fb8296e21b428.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,61 @@
+
+        this._messageCount = 0;
+        
+        let consoleMessage = new WI.ConsoleMessage(WI.mainTarget, WI.ConsoleMessage.MessageSource.Other, WI.ConsoleMessage.MessageLevel.Error, WI.UIString("Source Map loading errors"), WI.ConsoleMessage.MessageType.StartGroupCollapsed);
```
NIT: `consoleMessage` -> `titleConsoleMessage` to clarify the relationship with  `_titleConsoleMessageView`.

Web Inspector source doesn't have many code comments (the code is supposed to be clear enough to explain itself), but I feel that the behavior here warrants a short comment to explain the need for creating a mock console message + console message view (i.e.: a console group requires and initial message that acts as its title)

jdatapple replied:
Comment added and renamed the symbol for additional clarity in deba5de495d803c6f9703282319fb8296e21b428.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -293,6 +297,17 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
         WI.objectStores.consoleSnippets.putObject(snippet);
     }
 
+    _maybeAddConsoleMessageToConsoleSourceMapErrorGroup(message)
```
`_maybe` doesn't inspire a lot of confidence :)

This method name is a bit misleading because the messages aren't added to a source map error group (that happens elsewhere in the  `WI.JavaScritpLogViewController`). Here, they're just aggregated.

Perhaps a more descriptive name is `_collectFailedSourceMapConsoleMessage()` or similar.
The "maybe" is implied from the `if` conditions inside.

jdatapple replied:
I did not know the convention to use in a case where a decision like this was being made, but `collect` does seem more precise than `maybe`.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +369,35 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
                 this._currentSessionOrGroup.append(groupElement);
                 this._currentSessionOrGroup = group;
-            } else
+            } else {
```
Style: no brackets needed around single expression in `else` branch.

jdatapple replied:
Removed in deba5de495d803c6f9703282319fb8296e21b428.

rcaliman-apple reviewed (changes requested):
This looks even cleaner and clearer! 👏 

I left a note about an exception thrown for prematurely accessing `WI.ConsoleGroup.element`, a suggestion about how to reuse the error level styling by augmenting `WI.ConsoleGroup`, plus a few style comments and nits.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.css:
```diff
@@ -220,6 +230,18 @@
 
+.console-group-title > .console-error-level {
+    display: inline-block;
+    margin-right: 4px;
```
NIT: `margin-right` -> `margin-inline-end`

rcaliman-apple reviewed (approved):
Nice work! This looks clean and contained. 👍 Thank you for iterating on it!

Pinging @patrickangle and @dcrousso to review/approve as Reviewers.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js:
```diff
@@ -0,0 +1,56 @@
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL APPLE OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+garb * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
```
Oops?

jdatapple replied:
Addressed in 348f2f49dc517b2abe197f00972bf70aa92db4d8.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.css:
```diff
@@ -158,11 +158,21 @@
+}
+
+.console-group::before {
+    top: 4px; // because .console-group does not have a top/bottom padding like .console-message
```
Comments in CSS need to be wrapped in `/* */`, even for single-line comments. (`//` is not a valid comment in CSS)

jdatapple replied:
Addressed in 348f2f49dc517b2abe197f00972bf70aa92db4d8.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -43,6 +43,7 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
 
         this._snippets = new Set;
         this._restoringSnippets = false;
+        this._failedSourceMapConsoleMessages = new Set;
```
NIT: could you add a newline above this so that it's distinct from the snippets "section"?

jdatapple replied:
Addressed in e2c3874b5b35844fbcd028d68867b71662ae988e.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -293,6 +297,17 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
         WI.objectStores.consoleSnippets.putObject(snippet);
     }
 
+    _collectFailedSourceMapConsoleMessage(message)
```
NIT: can you move this to right after `_clearMessages` so that it's not in the middle of all the event handlers?

jdatapple replied:
Addressed in e2c3874b5b35844fbcd028d68867b71662ae988e.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -162,6 +164,8 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
             let issue = new WI.IssueMessage(message);
             this._issues.push(issue);
 
+            this._collectFailedSourceMapConsoleMessage(message);
```
please move this to the end of the `if` since it has nothing to do with the `issue` that's created above and used below

jdatapple replied:
Addressed in e2c3874b5b35844fbcd028d68867b71662ae988e.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -293,6 +297,17 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
+        if (!WI.settings.experimentalGroupSourceMapErrors.value)
+            return;
+
+        let isNetworkError = message.source === WI.ConsoleMessage.MessageSource.Network && message.level === WI.ConsoleMessage.MessageLevel.Error;
```
id turn this into an early return instead so that we dont have to call `WI.networkManager.isSourceMapURL` on every single console message

jdatapple replied:
Addressed in e2c3874b5b35844fbcd028d68867b71662ae988e.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +369,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
                 this._currentSessionOrGroup = group;
-            } else
-                this._currentSessionOrGroup.addMessageView(messageView);
+            } else this._currentSessionOrGroup.addMessageView(messageView);
```
Style: this should still be on a separate line

jdatapple replied:
Addressed in e2c3874b5b35844fbcd028d68867b71662ae988e.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +369,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+    _addConsoleMessageToConsoleSourceMapErrorGroup(messageView)
+    {
+            
+        if (this._firstSourceMapErrorMessageView === undefined) {
```
Style: please just check `if (!this._firstSourceMapErrorMessageView)` as we rarely like to distinguish between falsy and `undefined`

jdatapple replied:
Addressed in e2c3874b5b35844fbcd028d68867b71662ae988e.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +369,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+    
+    _addConsoleMessageToConsoleSourceMapErrorGroup(messageView)
+    {
+            
```
Style: extra newline, please remove

dcrousso replied:
FYI im still seeing this newline

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +369,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+            return;
+        }
+        
+        if (this._failedSourceMapsGroup === undefined) {
```
ditto

jdatapple replied:
Addressed in e2c3874b5b35844fbcd028d68867b71662ae988e.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +369,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+        
+        if (this._failedSourceMapsGroup === undefined) {
+            let currentSession = this._sessions.lastValue;
+            this._failedSourceMapsGroup = new WI.ConsoleSourceMapMessageGroup();
```
Style: no need for `()` if there are no arguments when constructing using `new`

jdatapple replied:
Addressed in e2c3874b5b35844fbcd028d68867b71662ae988e.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
```diff
@@ -229,6 +230,14 @@ WI.NetworkManager = class NetworkManager extends WI.Object
     get localResourceOverrides() { return this._localResourceOverrides; }
     get bootstrapScript() { return this._bootstrapScript; }
+    
+    get failedSourceMapURLs() {
```
Style: since these are simple `get` (i.e. they return a member variable of the same name) you can inline them at the end of the inlined `get` above
```
get bootstrapScript() { return this._bootstrapScript; }
get failedSourceMapURLs() { return this._failedSourceMapURLs; }
get downloadingSourceMaps() { return this._downloadingSourceMaps; }
```

jdatapple replied:
Addressed in e2c3874b5b35844fbcd028d68867b71662ae988e.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ConsoleGroup.js:
```diff
@@ -34,10 +34,12 @@ WI.ConsoleGroup = class ConsoleGroup extends WI.Object
         super();
 
         this._parentGroup = parentGroup;
+        this._element = false;
```
i think this should start out as `null` since it's going to be an object later not a boolean

jdatapple replied:
Addressed in e2c3874b5b35844fbcd028d68867b71662ae988e.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ConsoleGroup.js:
```diff
@@ -34,10 +34,12 @@ WI.ConsoleGroup = class ConsoleGroup extends WI.Object
     }
 
     // Public
-
```
Style: please keep this newline

jdatapple replied:
Addressed in e2c3874b5b35844fbcd028d68867b71662ae988e.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ConsoleGroup.js:
```diff
@@ -79,6 +80,35 @@ WI.ConsoleGroup = class ConsoleGroup extends WI.Object
         this._messagesElement.appendChild(messageOrGroupElement);
     }
+    
+    setMessageLevel(messageLevel)
```
Followup: it would be awesome to merge this logic with `WI.ConsoleMessageView.prototype.render` so that we don't have this code duplicated in two places :)

jdatapple replied:
This is an excellent point.

jdatapple replied:
Added a FIXME with a link to https://webkit.org/b/264489 in 503e2a6cacc0b56c26aaa3bdfcacb6fc6d55e6ae.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +369,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+        }
+        
+        if (this._failedSourceMapsGroup === undefined) {
+            let currentSession = this._sessions.lastValue;
```
why not `this._currentSessionOrGroup`?

jdatapple replied:
This is an intentional decision to avoid the `ConsoleSourceMapMessageGroup` from being added to an opened ConsoleGroup. As a special case of grouped errors, we want to ensure it is always top-level and not unintentionally hidden by being nested in a collapsed group whether created by us or by the user. cc @rcaliman-apple

rcaliman-apple replied:
> As a special case of grouped errors

These are resource loading errors that pertain to the main frame (and subframes) so we don't want to have them accidentally hidden inside an unrelated `ConsoleGroup`.

dcrousso replied:
i see.  so you're specifically trying to get the current top-level group.  in that case i'd rename this to `currentTopLevelSession` or something like that to better indicate the distinction

dcrousso replied:
FYI im still seeing the name as `currentSession`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -369,13 +369,33 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+            } else this._currentSessionOrGroup.addMessageView(messageView);
+            
+            if (WI.settings.experimentalGroupSourceMapErrors.value && WI.consoleManager.failedSourceMapConsoleMessages.has(messageView.message))
+                this._addConsoleMessageToConsoleSourceMapErrorGroup(messageView);
```
it's a bit odd that we do this after the above logic as i think that means the `messageView` gets `append` twice, once to the `_currentSessionOrGroup` and again to the `this._failedSourceMapsGroup`.  is it necessary for us to do that twice or could we move this check to be the primary `if` condition (with the existing `if` becoming `else if`)?

jdatapple replied:
This is part of how we handle allowing a single SourceMap error to be added as a normal error message, but after multiple SourceMap loading errors, we move them into the group. Are there enough performance concerns to warrant re-thinking this approach?

dcrousso replied:
it's probably not a big deal since it only applies to sourcemap errors (which are likely few/none of all the console messages), but it would be nice for this to not be a concern at all in the first place

i think we could have `_addConsoleMessageToConsoleSourceMapErrorGroup` return `true`/`false` based on whether it "consumed" the `messageView` (which is based on whether it's the first sourcemap error) and then if `true` it can early return at this callsite

im fine with this being a followup if you'd prefer :)

jdatapple replied:
Sounds like a plan.

jdatapple replied:
Filed https://webkit.org/b/264490

dcrousso replied:
can you please put a `// FIXME: <https://webkit.org/b/264490> (Improve failed SourceMap error grouping to avoid unnecessary work)` comment here so there's better linkage between bugzilla and the source code?

dcrousso reviewed (commented):
overall looks good!

a couple things id like to see addressed:

- make sure that any potentially expensive logic (e.g. checking for something in a `Set`) is done as late and last as possible to avoid any performance hits from having to do that to _every_ console message

- avoid mutating the DOM more than minimally necessary

- some style things

please let me know if anything is confusing or you'd like to discuss further :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
```diff
@@ -229,6 +230,8 @@ WI.NetworkManager = class NetworkManager extends WI.Object
     get localResourceOverrides() { return this._localResourceOverrides; }
     get bootstrapScript() { return this._bootstrapScript; }
+    get failedSourceMapURLs() { return this._failedSourceMapURLs; }
+    get downloadingSourceMaps() { return this._downloadingSourceMaps; }
```
it doesnt seem like these are actually used anywhere.  if so, please remove

jdatapple replied:
Added the `FIXEME`, removed the newline, changed `currentSession` to `currentTopLevelSession`, and removed the unnecessary getters in [7a17be1](https://github.com/WebKit/WebKit/pull/18733/commits/7a17be172e16a7e1cf40600f855ef61afda6465f).

jdatapple replied:
The getters were leftovers from an earlier exploration and didn't get cleaned up. Thanks for catching those!

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
```diff
@@ -1561,6 +1569,7 @@ WI.NetworkManager = class NetworkManager extends WI.Object
     _sourceMapLoadAndParseFailed(sourceMapURL)
     {
         this._downloadingSourceMaps.delete(sourceMapURL);
+        this._failedSourceMapURLs.add(sourceMapURL);
```
i think there should also be a `this._failedSourceMapURLs.clear();` in `_handleFrameMainResourceDidChange`, as otherwise we will always consider any URL within to be a failed sourcemap load even if it does succeed after the inspected page navigates (e.g. refreshing the page after connecting to VPN, adjust the inspected page URL to include a secret key, etc.)

jdatapple replied:
Agreed and added in 7a17be172e16a7e1cf40600f855ef61afda6465f.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ConsoleGroup.js:
```diff
@@ -79,6 +81,36 @@ WI.ConsoleGroup = class ConsoleGroup extends WI.Object
+        if (!this._element)
+            return;
+  
+        switch (messageLevel) {
```
do we need to cache `messageLevel` in case this is called _before_ `this._element` is initialized (i.e. this would automatically get called after `this._element` is created)?  or is it always expected that this will be called after (which is what it seems like this PR does)?

if the latter, we should replace the early return `if` above to just be a `console.assert(this._element);`

jdatapple replied:
It is designed to be called after `this._element` is initialized.

jdatapple replied:
I added the `console.assert` in 7a17be172e16a7e1cf40600f855ef61afda6465f.

dcrousso reviewed (commented):
i think just a few more fixes and we'll be good to go 😁

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ConsoleGroup.js:
```diff
@@ -79,6 +81,36 @@ WI.ConsoleGroup = class ConsoleGroup extends WI.Object
+    setMessageLevel(messageLevel)
+    {
+        if (!this._element)
+            console.assert(this._element);
```
sorry i wasnt entirely clear.  i was suggesting we get rid of the `if` as well

in this case we actually _have_ to get rid of the `if` as `console.*` are stripped from production builds, meaning we'd have a dangling `if`

```
setMessageLevel(messageLevel)
{
    console.assert(this._element);

    switch (messageLevel) {
    ...
    }
}
```

jdatapple replied:
TIL! Fixed in b59a15fd7276bb164a3b618611ab4ce7b6009788.

Thanks for all of the feedback. I learned a lot in the process!

dcrousso reviewed (approved):
awesome work!  thanks for iterating with me :)

one last thing to fix before merging, but otherwise great work!

jdatapple committed:
Add support for grouping source map loading errors.
https://bugs.webkit.org/show_bug.cgi?id=256683
rdar://problem/109239646

Reviewed by Devin Rousso.

This feature is behind an experimental Web Inspector setting
available in the Settings tab, under Experimental. When enabled and
more than one source map loading error occurs, network error messages
with a URL that matches a known source map URL are collected into an
expandable group with a counter showing the number of messages.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager):
(WI.ConsoleManager.prototype.get failedSourceMapConsoleMessages):
(WI.ConsoleManager.prototype.messageWasAdded):
(WI.ConsoleManager.prototype._handleSourceMapErrorMessage):
* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
(WI.JavaScriptLogViewController):
(WI.JavaScriptLogViewController.prototype.startNewSession):
(WI.JavaScriptLogViewController.prototype._didRenderConsoleMessageView):
(WI.JavaScriptLogViewController.prototype._handleFailedSourceMapMessageGrouping):
* Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
(WI.NetworkManager.prototype.get failedSourceMapURLs):
(WI.NetworkManager.prototype.get downloadingSourceMaps):
(WI.NetworkManager.prototype.isDownloadingSourceMapURL):
(WI.NetworkManager.prototype._sourceMapLoadAndParseFailed):
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.css:
(.console-sourcemap-errors-group .errors-count,):
(.console-message .console-sourcemap-failures-container,):
(.console-message.expandable.expanded .console-sourcemap-failures-container,):
(.console-message .repeat-count): Deleted.
(.console-message .console-message-extra-parameters-container,): Deleted.
(.console-message.expandable.expanded .console-message-extra-parameters-container,): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ConsoleSourceMapMessageGroup.js: Added.
(WI.ConsoleSourceMapMessageGroup):
(WI.ConsoleSourceMapMessageGroup.prototype.render):
(WI.ConsoleSourceMapMessageGroup.prototype.addMessageView):
(WI.ConsoleSourceMapMessageGroup.prototype.toggle):
(WI.ConsoleSourceMapMessageGroup.prototype._renderMessageCount):
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:

Canonical link: https://commits.webkit.org/270834@main
